code/modules/client/preference_setup/controls/01_keybindings.dm

```diff
@@ -0,0 +1,254 @@
+
+	var/list/user_binds = list()
+	for (var/key in key_bindings)
+		for(var/kb_name in key_bindings[key])
+			user_binds[kb_name] += list(key)
+	var/list/notadded = list()
+	for (var/name in global.keybindings_by_name)
+		var/datum/keybinding/kb = global.keybindings_by_name[name]
+		if(length(user_binds[kb.name]))
+			continue // key is unbound and or bound to something
+		var/addedbind = FALSE
+		if(hotkeys)
+			for(var/hotkeytobind in kb.hotkey_keys)
+				if(!length(key_bindings[hotkeytobind]) || hotkeytobind == "Unbound") //Only bind to the key if nothing else is bound expect for Unbound
+					LAZYADD(key_bindings[hotkeytobind], kb.name)
+					addedbind = TRUE
+		else
+			for(var/classickeytobind in kb.classic_keys)
+				if(!length(key_bindings[classickeytobind]) || classickeytobind == "Unbound") //Only bind to the key if nothing else is bound expect for Unbound
+					LAZYADD(key_bindings[classickeytobind], kb.name)
+					addedbind = TRUE
+		if(!addedbind)
+			notadded += kb
+
+	if(length(notadded))
+		addtimer(CALLBACK(src, .proc/announce_conflict, notadded), 5 SECONDS)
+
+/datum/preferences/proc/announce_conflict(list/notadded)
+	to_chat(client, "<span class='userdanger'>KEYBINDING CONFLICT!!!\n\
+	There are new keybindings that have defaults bound to keys you already set, They will default to Unbound. You can bind them in Setup Character or Game Preferences\n\
+	<a href='?src=\ref[src];preference=tab;tab=3'>Or you can click here to go straight to the keybindings page</a></span>")
+	for(var/item in notadded)
+		var/datum/keybinding/conflicted = item
+		to_chat(client, "<span class='userdanger'>[conflicted.category]: [conflicted.full_name] needs updating</span>")
+		LAZYADD(key_bindings["None"], conflicted.name) // set it to unbound to prevent this from opening up again in the future
+
+/datum/category_item/player_setup_item/controls/keybindings
+	name = "Keybindings"
+	sort_order = 1
+
+/datum/category_item/player_setup_item/controls/keybindings/load_preferences(datum/pref_record_reader/R)
+	pref.key_bindings = R.read("key_bindings")
+
+/datum/category_item/player_setup_item/controls/keybindings/sanitize_preferences()
+	pref.key_bindings = sanitize_keybindings(pref.key_bindings)
+	pref.check_keybindings()
+
+/datum/category_item/player_setup_item/controls/keybindings/save_preferences(datum/pref_record_writer/W)
+	W.write("key_bindings", pref.key_bindings)
+
+/datum/category_item/player_setup_item/controls/keybindings/content(mob/user)
+	. = list()
+	// Create an inverted list of keybindings -> key
+	var/list/user_binds = list()
+	for (var/key in pref.key_bindings)
+		for(var/kb_name in pref.key_bindings[key])
+			user_binds[kb_name] += list(key)
+
+	var/list/kb_categories = list()
+	// Group keybinds by category
+	for (var/name in global.keybindings_by_name)
+		var/datum/keybinding/kb = global.keybindings_by_name[name]
+		kb_categories[kb.category] += list(kb)
```

Comment:
Why are we doing this stuff each time the UI is refreshed?

Reply:
Responding here because I can't respond elsewhere, but ironically a bunch of your comments can't be replied to.

Reply:
I didn't come with UI part very well. It in state of Just WorksTM.